src/billing_handler.rs

```diff
@@ -37,47 +78,60 @@ pub async fn sign_data(appstate: Data<AppState>, data: Json<SigningData>) -> imp
             .body(format!("Error decoding nonce into 32 bytes: {}", err));
     }
 
-    let mut signed_data: Vec<u8> = bytes32_nonce.to_vec();
-    for txhash in signing_data.txhashes {
-        if let Some(cost) = appstate.execution_costs.lock().unwrap().remove(&txhash) {
-            let mut bytes32_txhash = [0u8; 32];
-            if let Err(err) = hex::decode_to_slice(&txhash[2..], &mut bytes32_txhash) {
-                return HttpResponse::InternalServerError().body(format!(
-                    "Error decoding transaction hash into 32 bytes: {}",
-                    err
-                ));
+    let mut bill_claim_data = bytes32_nonce.to_vec();
+    for tx_hash in signing_data.tx_hashes {
+        if let Some(cost) = appstate.execution_costs.lock().unwrap().remove(&tx_hash) {
+            let mut bytes32_tx_hash = [0u8; 32];
+            if let Err(_) = hex::decode_to_slice(&tx_hash[2..], &mut bytes32_tx_hash) {
+                continue;
             }
 
-            signed_data.append(&mut bytes32_txhash.to_vec());
-            signed_data.append(&mut cost.to_be_bytes().to_vec());
-        } else {
-            return HttpResponse::BadRequest().body(format!(
-                "{} tx hash doesn't exist in the current bill",
-                txhash
-            ));
+            bill_claim_data.append(&mut bytes32_tx_hash.to_vec());
+            bill_claim_data.append(&mut cost.to_be_bytes().to_vec());
         }
     }
 
-    let mut hasher = Keccak::v256();
-    hasher.update(&signed_data);
+    if bill_claim_data.len() == 32 {
+        return HttpResponse::BadRequest().body("No tx present in the bill");
+    }
 
-    let mut hash = [0u8; 32];
-    hasher.finalize(&mut hash);
+    let signature = sign_data(bill_claim_data.as_slice(), &appstate.signer).await;
+    if signature.is_err() {
```

Comment:
If signature is failing here, then it will fail again when `billing/latest` will be called. There is no change in parameters or other context data. Signature failing here seems fatal error to me. Double check where it can fail and how developers has suggested to handle those errors.

Reply:
So, if signing is throwing exception then it will be there for every sign call. That means this application is not usable at all in that case, in production.

Then I think, if you just return exception here, that would be sufficient. Any exception due to implementation change, would be hit during testing phase of release.

You can remove the special handling here, it looks like you are trying to recover from error, but it's not. Kinda misleading.

Reply:
I was just handling it the way it was done when the serverless requests are getting processed by the enclave. So shouldn't we add an exception here as well:- 

  pub async fn get_workerd_response(

    port: u16,
    req: HttpRequest,
    body: actix_web::web::Bytes,
    signer: &k256::ecdsa::SigningKey,
    host_header: &str,

  ) -> Result<HttpResponse, anyhow::Error> {

    let mut hasher = Keccak::v256();
    hasher.update(b"|oyster-serverless-hasher|");

    let timestamp = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)?
        .as_secs();
    hasher.update(b"|timestamp|");
    hasher.update(&timestamp.to_be_bytes());

    hasher.update(b"|request|");
    hasher.update(b"|method|");
    hasher.update(req.method().to_string().as_bytes());
    hasher.update(b"|pathandquery|");
    hasher.update(
        req.uri()
            .path_and_query()
            .map(|x| x.as_str())
            .unwrap_or("")
            .as_bytes(),
    );
    hasher.update(b"|host|");
    hasher.update(host_header.as_bytes());
    hasher.update(b"|body|");
    hasher.update(&body);

    let port_str = port.to_string();
    let req_url = "http://127.0.0.1:".to_string() + &port_str + "/";
    let client = reqwest::Client::new();
    let response = req
        .headers()
        .into_iter()
        .fold(
            client.request(req.method().clone(), req_url),
            |req, header| req.header(header.0.clone(), header.1.clone()),
        )
        .body(body)
        .send()
        .await?;
    hasher.update(b"|response|");

    let mut actix_resp = response.headers().into_iter().fold(
        HttpResponse::build(response.status()),
        |mut resp, header| {
            resp.append_header((header.0.clone(), header.1.clone()));
            resp
        },
    );
    let response_body = response.bytes().await?;

    hasher.update(b"|body|");
    hasher.update(&response_body);

    let mut hash = [0u8; 32];
    hasher.finalize(&mut hash);

    let (rs, v) = signer.sign_prehash_recoverable(&hash)?;`

Reply:
Here in this case `?` is used to pass on exception. No explicit handling required. Even for your case, I was suggesting just respond back with InternalError. And that would be fine.

Reply:
> If signature is failing here, then it will fail again when `billing/latest` will be called. There is no change in parameters or other context data. Signature failing here seems fatal error to me. Double check where it can fail and how developers has suggested to handle those errors.

Depends on why it can fail and if it's consistent or not. For example, pretty sure it generates a random number inside, which can probably fail inconsistently.

Reply:
> Btw If we look more closely, there is an aggressive exit happening when serving the requests as well after failing to sign the requests/response data. So even after getting a successful response from the worker serving the app, we're throwing an error response to the user of the app. Is this an acceptable behaviour? @roshanr95 @vg-27

Yes, responses are insecure without a signature.

Reply:
> For example, pretty sure it generates a random number inside, which can probably fail inconsistently.

Yeah, that's right. Plain ECDSA signature uses random number. But here [implementation](https://docs.rs/ecdsa/0.16.9/src/ecdsa/recovery.rs.html#180) seems to be using [RFC6979](https://www.rfc-editor.org/rfc/rfc6979.html), where `k` is computed deterministically.

Anyways, if there is possibility of inconsistent error. Let's keep the backup of billing  data.